Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -21,981 +21,660 @@
-
-    return (CGImageRef)[img CGImage];
-}
+CFTypeID CGImageGetTypeID() { return __CGImage::GetTypeID(); }
 
 /**
  @Status Caveat
- @Notes Doesn't support copy-on-write semantics - returns an unlinked copy of the source
-        image cropped to the specified rectangle.
+ @Notes Limited bitmap formats available and decode is not supported.
 */
-CGImageRef CGImageCreateWithImageInRect(CGImageRef ref, CGRect rect) {
-    if (ref == NULL) {
-        TraceWarning(TAG, L"CGImageCreateWithImageInRect: ref = NULL!");
-        return 0;
-    }
-
-    rect = CGRectIntegral(rect);
+CGImageRef CGImageCreate(size_t width, size_t height, size_t bitsPerComponent,
+                         size_t bitsPerPixel, size_t bytesPerRow,
+                         CGColorSpaceRef colorSpace, CGBitmapInfo bitmapInfo,
+                         CGDataProviderRef provider, const float *decode,
+                         bool shouldInterpolate,
+                         CGColorRenderingIntent intent) {
 
-    CGRect imgRefSize;
-    imgRefSize.origin.x = 0;
-    imgRefSize.origin.y = 0;
-    imgRefSize.size.width = (float)ref->Backing()->Width();
-    imgRefSize.size.height = (float)ref->Backing()->Height();
+  RETURN_NULL_IF(((provider == nullptr) ||
+                  ![(NSObject *)provider isKindOfClass:[NSData class]]));
 
-    rect = CGRectIntersection(rect, imgRefSize);
+  NSData *dataProvider = (__bridge NSData *)provider;
 
-    __CGSurfaceInfo surfaceInfo;
-    ref->Backing()->GetSurfaceInfoWithoutPixelPtr(&surfaceInfo);
+  unsigned char *data = (unsigned char *)[dataProvider bytes];
 
-    // Override width and height with the rect
-    surfaceInfo.width = rect.size.width;
-    surfaceInfo.height = rect.size.height;
+  bool colorSpaceAllocated = false;
+  if (colorSpace == NULL) {
+    if (bytesPerRow >= (width * 3)) {
+      TraceWarning(
+          TAG,
+          L"Warning: colorSpace = NULL, assuming RGB based on bytesPerRow.");
+      colorSpace = CGColorSpaceCreateDeviceRGB();
+    } else {
+      TraceWarning(
+          TAG,
+          L"Warning: colorSpace = NULL, assuming Gray based on bytesPerRow.");
+      colorSpace = CGColorSpaceCreateDeviceGray();
+    }
 
-    assert(surfaceInfo.surfaceData == NULL);
+    colorSpaceAllocated = true;
+  }
 
-    CGImageRef newImage = new CGBitmapImage(surfaceInfo);
+  ComPtr<IWICBitmap> image;
+  ComPtr<IWICImagingFactory> imageFactory = _GetWICFactory();
 
-    int startX = (int)rect.origin.x;
-    int startY = (int)rect.origin.y;
-    int sizeX = (int)rect.size.width;
-    int sizeY = (int)rect.size.height;
+  REFGUID pixelFormat = _CGImageGetWICPixelFormat(
+      bitsPerComponent, bitsPerPixel, colorSpace, bitmapInfo);
 
-    BYTE* srcIn =
-        ((BYTE*)ref->Backing()->LockImageData()) + startY * ref->Backing()->BytesPerRow() + startX * ref->Backing()->BytesPerPixel();
-    BYTE* destOut = (BYTE*)newImage->Backing()->LockImageData();
+  // TODO: take care of the stride (mapping of the guids)
```

Comment:
Taking care of it in the next CR iteration.

---
In reply to: [84331167](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84331167) [](ancestors = 84331167)